sdk/cosmosdb/cosmos/src/CosmosDiagnostics.ts

```diff
@@ -0,0 +1,146 @@
+  }
+
+  public readonly clientSideRequestStatistics: ClientSideRequestStatistics;
+
+  public get getContactedRegionNames(): Set<string> {
+    const locations = this.cosmosDiagnosticContext.locationEndpointsContacted.values();
+    return new Set([...locations].map((location) => location.name));
+  }
+}
+
+/**
+ * This type contains diagnostic information regarding all metadata request to server during an CosmosDB client operation.
+ */
+export type MetadataLookUpDiagnostics = {
+  metadataLookups: MetadataLookUpDiagnostic[];
+};
+
+/**
+ * This type captures diagnostic information regarding retries attempt during an CosmosDB client operation.
+ */
+export type RetryDiagnostics = {
+  failedAttempts: FailedRequestAttemptDiagnostic[];
+};
+
+/**
+ * This type contains diagnostic information regarding a single metadata request to server.
+ */
+export interface MetadataLookUpDiagnostic {
+  id: string;
+  startTimeUTCInMs: number;
+  endTimeUTCInMs: number;
+  metaDataType: MetadataLookUpType;
+  activityId: string;
+}
+
+/**
+ * This type captures diagnostic information regarding a failed request to server api.
+ */
+export interface FailedRequestAttemptDiagnostic {
+  id: string;
+  attemptNumber: number;
+  startTimeUTCInMs: number;
+  endTimeUTCInMs: number;
+  statusCode: string;
+}
+
+/**
+ * This is enum for Type of Metadata lookups possible.
+ */
+export enum MetadataLookUpType {
+  PartitionKeyRangeLookUp = "PARTITION_KEY_RANGE_LOOK_UP",
+  ServerAddressLookUp = "SERVER_ADDRESS_LOOK_UP",
+  DatabaseAccountLookUp = "DATABASE_ACCOUNT_LOOK_UP",
+}
+
+/**
+ * This is a collection type for all client side diagnostic information.
+ */
+export type ClientSideRequestStatistics = {
+  /**
+   * This is the UTC timestamp for start of client operation.
+   */
+  requestStartTimeUTCInMs: number;
+  /**
+   * This is the UTC timestamp for end of client operation.
+   */
+  requestEndTimeUTCInMs: number;
+  /**
+   * This is the activityId for request, made to server for fetching the requested resource. (As opposed to other potential meta data requests)
+   */
+  activityId: string;
+  /**
+   * This is the list of Location Endpoints contacted during the client operation.
+   */
+  locationEndpointsContacted: Location[];
+  /**
+   * This field captures diagnostic information for retries happened during client operation.
+   */
+  retryDiagnostics: RetryDiagnostics;
+  /**
+   * This field captures diagnostic information for meta data lookups happened during client operation.
+   */
+  metadataDiagnostics: MetadataLookUpDiagnostics;
+  /**
+   * This is the payload length, in bytes made to server for the client operation requested. (This doesn't include payloads for meta data requests)
+   */
+  requestPayloadLength: number;
+  /**
+   * This is the payload length, in bytes made to recieved from server for the client operation requested. (This doesn't include payloads for meta data responses)
```

Comment:
made to recieved from server --> received from server

Reply:
Ack